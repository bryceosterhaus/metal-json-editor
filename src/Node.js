import Component, {Config} from 'metal-jsx';

import Tree from './Tree';
import EditableValue from './EditableValue';

class Node extends Component {
	getBracket(value, open = true) {
		const array = Array.isArray(value);

		let retVal = open ? '{' : '}';

		if (array) {
			retVal = open ? '[' : ']';
		}

		return retVal;
	}

	getSize(value) {
		const array = Array.isArray(value);

		return array ? value.length : Object.keys(value).length
	}

	handleExpand() {
		this.props.onToggleExpand(this.props.value);
	}

	render() {
		const {
			arrowRenderer,
			expandedNodes,
			locator,
			name,
			onChange,
			value
		} = this.props;

		const expanded = expandedNodes.has(value);

		const readOnly = value !== undefined && value !== null && value.__metal_devtools_read_only;

		const displayAsObj = value instanceof Object && !readOnly;

		return (
			<li>
				<div>
					<span onClick={this.handleExpand.bind(this)} style={displayAsObj ? 'cursor: pointer;' : ''}>
						{`${name}: `}
					</span>

					{!displayAsObj &&
						<EditableValue
							locator={locator}
							name={name}
							onChange={onChange}
							value={value}
							readOnly={readOnly}
						/>
					}

					{displayAsObj &&
						<span>
							<span onClick={this.handleExpand.bind(this)} style="cursor: pointer;">
								{arrowRenderer(expanded)}
							</span>

							{this.getBracket(value)}

							{!expanded &&
								this.getSize(value)
							}

							{!expanded &&
								this.getBracket(value, false)
							}
						</span>
					}

					{expanded && displayAsObj &&
						<Tree
							arrowRenderer={arrowRenderer}
							data={value}
							expandedNodes={expandedNodes}
							locator={locator}
							onChange={onChange}
							onToggleExpand={this.props.onToggleExpand}
						/>
					}

					{expanded && displayAsObj &&
						this.getBracket(value, false)
					}
				</div>
			</li>
		);
	}
}

Node.PROPS = {
	arrowRenderer: Config.func().value(() => {}),
	expandedNodes: Config.instanceOf(WeakSet).required(),
	locator: Config.array().value([]),
	name: Config.string(),
	onChange: Config.func(),
	onToggleExpand: Config.func(),
	value: Config.any()
};

export default Node;
