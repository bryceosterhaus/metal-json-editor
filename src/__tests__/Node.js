import Node from '../Node';

describe('Node', () => {
  let component;

  afterEach(() => {
    if (component) {
      component.dispose();
    }
  });

  it('renders', () => {
    component = new Node({expandedNodes: new WeakSet()});

    expect(component).toMatchSnapshot();
  });

  it('should return the proper bracket', () => {
    component = new Node({expandedNodes: new WeakSet()});

    expect(component.getBracket([])).toMatchSnapshot();
    expect(component.getBracket([], false)).toMatchSnapshot();

    expect(component.getBracket({})).toMatchSnapshot();
    expect(component.getBracket({}, false)).toMatchSnapshot();
  });

  it('should return the size of the object or array', () => {
    component = new Node({expandedNodes: new WeakSet()});

    expect(component.getSize({one: 1, two: 2})).toMatchSnapshot();
    expect(component.getSize([1, 2, 3])).toMatchSnapshot();
  });

  it('should call onToggleExpand prop', () => {
    const spy = jest.fn();
    const value = 123;

    component = new Node({
      expandedNodes: new WeakSet(),
      onToggleExpand: spy,
      value
    });

    component.handleExpand();

    expect(spy).toHaveBeenCalledWith(value);
  });
});
