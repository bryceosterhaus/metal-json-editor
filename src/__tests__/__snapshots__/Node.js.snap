// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node renders 1`] = `
<li>
  <div key="_metal_jsx_1">
    <span style="" key="_metal_jsx_1">undefined: </span>
    <span style="" title="Click to Edit" class="json-editor-value" key="_metal_jsx_2">
      <span key="_metal_jsx_2"> "" </span>
    </span>
  </div>
</li>
`;

exports[`Node should return the proper bracket 1`] = `"["`;

exports[`Node should return the proper bracket 2`] = `"]"`;

exports[`Node should return the proper bracket 3`] = `"{"`;

exports[`Node should return the proper bracket 4`] = `"}"`;

exports[`Node should return the size of the object or array 1`] = `2`;

exports[`Node should return the size of the object or array 2`] = `3`;
